Tuxedo Control Center

blitz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

[Tuxedo](https://www.tuxedocomputers.com/) is a German laptop manufacturer that provides Linux-friendly laptops. Their system control is done via an app called "Tuxedo Control Center" (TCC). This open source app provides fan control settings among other things. Without this app, the Tuxedo laptops default to very noisy fan control settings. It lives on [Github](https://github.com/tuxedocomputers/tuxedo-control-center).

This PR:
- adds the Tuxedo Control Center application and daemon, and
- a tuxedo-control-center module to enable the daemon at boot,
- bumps tuxedo-keyboard to its latest version.

Testing done:

I've been running TCC on a Tuxedo XP14 laptop without issues. Fan control notably improves compared to running it without TCC.

This PR is based to a large degree on work done by @Vodurden.

Fixes #132206.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

blitz committed:
tuxedo-control-center: init at 1.1.2

This is based off work from:
https://github.com/Vodurden/devbox/blob/f06f565286e069ec5c58c222c97e37f0f7c5b551/nix/modules/nixos/tuxedo-control-center.nix

Editing has taken place in https://github.com/blitz/tuxedo-nixos/ and
can be seen there.

blitz committed:
modules/tuxedo-control-center: init

Mic92 reviewed (changes requested):
Please integrate the node package into our `nodePackages` in `pkgs/development/node-packages` set. 12000 lines of code for a single package is too much.

blitz commented:
> Please integrate the node package into our `nodePackages` in `pkgs/development/node-packages` set. 12000 lines of code for a single package is too much.

@Mic92 So in general I agree. The problem is that TCC needs its `devDependencies` to build which are not included in `nodePackages`. This results in build failure:

```
Error: Cannot find module 'node-addon-api'
Require stack:
- /nix/store/xhnnim063mxkv7y2810yr75zn7j429r5-tuxedo-control-center-1.0.14/lib/node_modules/tuxedo-control-center/[eval]
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at [eval]:1:1
    at Script.runInThisContext (vm.js:134:12)
    at Object.runInThisContext (vm.js:310:38)
    at internal/process/execution.js:81:19
    at [eval]-wrapper:6:22
    at evalScript (internal/process/execution.js:80:60) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/xhnnim063mxkv7y2810yr75zn7j429r5-tuxedo-control-center-1.0.14/lib/node_modules/tuxedo-control-center/[eval]'
  ]
```

Other people seem to have this problem when packaging as well (see https://github.com/svanderburg/node2nix/issues/149), without any clear solution. Do you have a pointer?

blitz commented:
@Mic92 I've tried to integrate TCC into `nodePackages`, but there are two roadblocks:

- TCC needs dev dependencies to be avaiable to do its build, this could probably worked around by patching the `package.json`, but it feels very hacky.
- it seems that the way `nodePackages` is composed, it doesn't respect TCC's `package-lock.json`, so you end up with incompatible libraries / untested library combinations.

So unless I'm missing something obvious packaging, TCC apart from `nodePackages` seems to be the way to go. It seems I'm also not in uncharted territory here, because there are quite a few Node/Electron packages that go the same route:

```
% find . -name "node-packages.nix" | wc -l
25
```

Don't understand me wrong, I would be happy if integrating it into `nodePackages` works, but I really don't see how. :(

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/update.sh:
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Please use nix-shell with curl and node2nix.

blitz replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/update.sh:
```diff
@@ -0,0 +1,37 @@
+
+SCRIPTDIR="$( dirname "${BASH_SOURCE[0]}" )"
+NIXPKGS="$SCRIPTDIR"/../../../..
+node2nix="$(nix-build "$NIXPKGS" -A nodePackages.node2nix)/bin/node2nix"
+curl="$(nix-build "$NIXPKGS" -A curl)/bin/curl"
```
```suggestion
```

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,159 @@
+   '';
+
+  buildPhase = ''
+    set -x
```
```suggestion
```

blitz replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,159 @@
+    makeWrapper
+
+    # For node-gyp
+    python
```
```suggestion
    python2
```

blitz replied:
I've changed this to use python3, which works fine as well. No reason to add more dependencies on python2.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,159 @@
+    cp $out/data/dist-data/tuxedo-control-center_256.svg \
+       $out/share/icons/hicolor/scalable/apps/tuxedo-control-center.svg
+
+    ${desktopItem.buildCommand}
```
```suggestion
    runHook postInstall
```

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,159 @@
+  '';
+
+  installPhase = ''
+    mkdir -p $out
```
```suggestion
    runHook preInstall

    mkdir -p $out
```

blitz replied:
I've added pre/post hooks to build and install.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,159 @@
+    NG_CLI_ANALYTICS="false";
+  };
+
+  desktopItem = makeDesktopItem {
```
Please use copyDesktopItems.

blitz replied:
Done.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/README.md:
```diff
@@ -0,0 +1,6 @@
+# How to update
+
+1. Bump the version of `tuxedo-control-center` in [packages.json](./packages.json) to the latest on [tuxedocomputers/tuxedo-control-center](https://github.com/tuxedocomputers/tuxedo-control-center)
+2. Run `generate-dependencies.sh`
+3. Set the version in [default.nix](./default.nix). (Version can be found in [tuxedocomputers/tuxedo-control-center/package.json](https://github.com/tuxedocomputers/tuxedo-control-center/blob/master/package.json))
+4. Build and test
```
Maybe move this inline next to version so that reviewer also see it.

blitz replied:
Done.

blitz commented:
@ofborg build tuxedo-control-center

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,179 @@
+{ pkgs, lib, stdenv, copyDesktopItems,
+
+  python3, udev,
+
+  makeWrapper, nodejs, electron_11, fetchFromGitHub
+}:
```
```suggestion
{ pkgs, lib, stdenv, copyDesktopItems
, python3, udev
. makeWrapper, nodejs, electron_11, fetchFromGitHub
}:
```

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,179 @@
+}:
+
+let
+  baseName = "tuxedo-control-center";
```
```suggestion
  pname = "tuxedo-control-center";
```

blitz replied:
Fixed.

SuperSandro2000 commented on pkgs/os-specific/linux/tuxedo-control-center/default.nix:
```diff
@@ -0,0 +1,179 @@
+in
+
+stdenv.mkDerivation rec {
+  name = "${baseName}-${version}";
```
```suggestion
  inherit pname version;
```

blitz replied:
Fixed.

blitz commented:
@SuperSandro2000 I've resolved the issues. Please have another look. :)

blitz commented:
I've rebased this and updated Tuxedo Control Center to 1.1.2.

MichaelAkvo commented:
Is there a way I can use this while waiting for it to get merged?

pgronkievitz commented:
> Is there a way I can use this while waiting for it to get merged?
> 
> P.S thanks a lot for the work!

@blitz has made a module (unfortunately it's not a flake, but I've managed to work around this)

https://github.com/blitz/tuxedo-nixos

blitz commented:
> > Is there a way I can use this while waiting for it to get merged?
> > 
> > P.S thanks a lot for the work!
> 
> @blitz has made a module (unfortunately it's not a flake, but I've managed to work around this)
> 
> https://github.com/blitz/tuxedo-nixos

I'm slowly flakifying my NixOS configs so this repo will also become a flake soon.

MichaelAkvo commented:
@blitz you the real MVP.

Thanks for pointing me to the repo @pgronkievitz

dasj19 commented:
@MichaelAkvo if you are using the `unstable` branch of nixpkgs you can use the PR author's branch as your base for nixpkgs.
You can clone his fork and switch to the tuxedo branch then run this:
`nixos-rebuild -I nixpkgs=/path/to/your/version/of/nixpkgs` switch

As for myself I have some migration work to do on my tuxedo computer then I'll start using this and come with feedback

dasj19 reviewed (commented):
There is a merge conflict and the following message
`error: Package ‘electron-11.5.0’ in /home/daniel/workspace/projects/nixpkgs/pkgs/development/tools/electron/generic.nix:25 is marked as insecure, refusing to evaluate.`

I added the electron package to the exceptions for now and started building.

Edit: build finished and got this result:
```

/bin/sh: line 1: prime-select: command not found
/bin/sh: line 1: prime-supported: command not found
(node:74258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
(node:74258) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'on' of undefined
    at TccDBusController.onModeReapplyPendingChanged (/nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/common/classes/TccDBusController.js:206:24)
    at /nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/e-app/main.js:150:17
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:74258) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:74258) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

blitz commented:
@dasj19 Thanks for testing!

> There is a merge conflict 

I've rebased the PR, so this should be gone. If you actually want to use it while it's not merged, I would suggest you head over to [this repo](https://github.com/blitz/tuxedo-nixos/).

> and the following message `error: Package ‘electron-11.5.0’ in /home/daniel/workspace/projects/nixpkgs/pkgs/development/tools/electron/generic.nix:25 is marked as insecure, refusing to evaluate.`

Checkout [blitz/tuxedo-nixos#troubleshooting](https://github.com/blitz/tuxedo-nixos/#troubleshooting) and https://github.com/tuxedocomputers/tuxedo-control-center/issues/148. 

> 
> I added the electron package to the exceptions for now and started building.
> 
> Edit: build finished and got this result:
> 
> ```
> 
> /bin/sh: line 1: prime-select: command not found
> /bin/sh: line 1: prime-supported: command not found

This is harmless, unless you have a nVidia PRIME system. In this case, please say so. It would be interesting to see whether we get this to work.

> (node:74258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
> (node:74258) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'on' of undefined
>     at TccDBusController.onModeReapplyPendingChanged (/nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/common/classes/TccDBusController.js:206:24)
>     at /nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/e-app/main.js:150:17
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
> (node:74258) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:74258) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ```

These need to be reported upstream. If there is specific broken functionality, we can try to patch it. I would welcome help for this.

blitz commented:
@pgronkievitz If you want to give the flake a shot, there is a flake branch here that exports a module: https://github.com/blitz/tuxedo-nixos/tree/flake

dasj19 commented:
@blitz Thanks for your response and fix.

I tried the new changes and it behaves as follows:
1. When I put `tuxedo-control-center` in `environment.systemPackages` in my configuration.nix it ends with the same error you said that should be sent upstream
2. when I build with `hardware.tuxedo-control-center.enable = true;` that error does not show up but instead  I get only:
```
/bin/sh: line 1: prime-select: command not found
/bin/sh: line 1: prime-supported: command not found
(node:37258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
```
And a blank Tuxedo Control Center shows up.

The same happens when I use https://github.com/blitz/tuxedo-nixos/#usage

My hardware us this https://www.tuxedocomputers.com/en/Linux-Hardware/Linux-Notebooks/15-16-inch/TUXEDO-Book-XA15-Gen10.tuxedo with max specs.

blitz commented:
> @blitz Thanks for your response and fix.
> 
> I tried the new changes and it behaves as follows:
> 
>     1. When I put `tuxedo-control-center` in `environment.systemPackages` in my configuration.nix it ends with the same error you said that should be sent upstream

Ah, I didn't realize that you don't use the module. Without the module, you don't get the TCC daemon and there is some dbus magic missing. So this is not expected to work.

> Edit: now it launches successfully using https://github.com/blitz/tuxedo-nixos/#usage I guess I had to kill the previous process

Yes, when you start TCC without the daemon and dbus magic it gets stuck in a weird state. I don't think there is anything we an do about it.

> Now a review of the app:

It would be interesting, if the limitations you see also happen on Ubuntu or TuxedoOS. If so, this is an upstream limitation.

dasj19 commented:
> It would be interesting, if the limitations you see also happen on Ubuntu or TuxedoOS. If so, this is an upstream limitation.
Hi, I figure that the prebuilt binaries for ubuntu 20.04 predates the machine I have and there is no point in trying them. So I tried compiling on ubuntu 22.04 live, but I can't figure out how to get tccd.service as mentioned in step 3. here https://github.com/tuxedocomputers/tuxedo-control-center#development-setup

Whithout the service the app crashes right after launch. So if you know of a `tccd.service` and `tccd-sleep.service` it will be handy. Otherwise I am unable to test on Ubuntu

blitz commented:
> So if you know of a tccd.service and tccd-sleep.service it will be handy.

Thanks for trying!

See here for the service files: https://github.com/tuxedocomputers/tuxedo-control-center/tree/master/src/dist-data

blitz commented:
> @blitz thanks for guiding me so far. I did not succeeded in having a working build from source. The resulted `tccd` binary gives segmentation fault.
> But I installed the deb packages provided by tuxedo for Ubuntu 20.04 on my Ubuntu 22.04 live usb.
> 
> The conclusion is that there is fan control support, and none of the features are unsupported. As I mentioned previously in NixOS there were certain parts of the application that printed "This feature is unsupported". In Ubuntu everything was supported.

Did you reboot after activating the module on NixOS? The module also loads a kernel module that is required for fan control etc.

dasj19 commented:
> Did you reboot after activating the module on NixOS? The module also loads a kernel module that is required for fan control etc.

I don't know what happened but after reboot the TCC displays a blank window. When started from terminal I get `TUXEDO Control Center is already running`. Killing the processes and trying again does not help.

journalctl says:
`Jan 27 07:55:31 tuxedo tccd[1445]: CpuWorker: Error validating/reapplying profile => Error: Could not read value from path: /sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq => Error: ENOENT: no such file or directory, open '/sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq => Error: ENOENT: no such file or directory, open '/sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq'`

blitz commented:
@dasj19 Having cpufreq sysfs files only for some CPUs is extremely weird. Do you have an unusual kernel configuration or any other special config?

That being said, given that this discussion is not necessarily regarding packaging, I would propose to continue the rootcausing as an issue in https://github.com/blitz/tuxedo-nixos/.

NANASHI0X74 commented:
what's the current status of this PR? anyone currently working on it? Is it only still blocked by the upstream electron issue?

blitz commented:
> what's the current status of this PR? anyone currently working on it? Is it only still blocked by the upstream electron issue?
> 

I'm still interested in getting this into Nixpkgs. But upstream using ancient Electron is still an issue. Otherwise there is the issue of checking in gigantic autogenerated Nix files.

blitz commented:
I'm closing this for now. If someone wants to revive it, feel free to checkout https://github.com/blitz/tuxedo-nixos/
